libcontainer/cgroups/fs2/memory.go

```diff
@@ -126,12 +146,93 @@ func getMemoryDataV2(path, name string) (cgroups.MemoryData, error) {
 
 	value, err = fscommon.GetCgroupParamUint(path, limit)
 	if err != nil {
-		if moduleName != "memory" && os.IsNotExist(err) {
-			return cgroups.MemoryData{}, nil
-		}
 		return cgroups.MemoryData{}, errors.Wrapf(err, "failed to parse %s", limit)
 	}
 	memoryData.Limit = value
 
 	return memoryData, nil
 }
+
+func statsFromMeminfo(stats *cgroups.Stats) error {
+	f, err := os.Open("/proc/meminfo")
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	// Fields we are interested in.
+	var (
+		swap_free        uint64
+		swap_total       uint64
+		main_total       uint64
+		main_available   uint64
+		main_free        uint64
+		main_buffers     uint64
+		page_cache       uint64
+		slab_reclaimable uint64
+	)
+	mem := map[string]*uint64{
+		"SwapFree":     &swap_free,
+		"SwapTotal":    &swap_total,
+		"MemTotal":     &main_total,
+		"MemAvailable": &main_available,
+		"MemFree":      &main_free,
+		"Buffers":      &main_buffers,
+		"Cached":       &page_cache,
+		"SReclaimable": &slab_reclaimable,
+	}
+
+	found := 0
+	sc := bufio.NewScanner(f)
+	for sc.Scan() {
+		parts := strings.SplitN(sc.Text(), ":", 3)
+		if len(parts) != 2 {
+			// Should not happen.
+			continue
+		}
+		k := parts[0]
+		p, ok := mem[k]
+		if !ok {
+			// Unknown field -- not interested.
+			continue
+		}
+		vStr := strings.TrimSpace(strings.TrimSuffix(parts[1], " kB"))
+		*p, err = strconv.ParseUint(vStr, 10, 64)
+		if err != nil {
+			return errors.Wrap(err, "parsing /proc/meminfo "+k)
+		}
+
+		found++
+		if found == len(mem) {
+			// Got everything we need -- skip the rest.
+			break
+		}
+	}
+	if sc.Err() != nil {
+		return sc.Err()
+	}
+
+	stats.MemoryStats.SwapUsage.Usage = (swap_total - swap_free) * 1024
+	stats.MemoryStats.SwapUsage.Limit = math.MaxUint64
+
+	// The code below is adapted from procps-ng free(1), see
+	// https://gitlab.com/procps-ng/procps/-/blob/e319650278/proc/sysinfo.c#L781-792
+
+	main_cached := page_cache + slab_reclaimable
+
+	// If main_available is greater than main_total or our calculation of
+	// mem_used overflows, that's symptomatic of running within a lxc
+	// container where such values will be dramatically distorted over
+	// those of the host.
+	if main_available > main_total {
+		main_available = main_free
+	}
+	mem_used := int64(main_total - main_free - main_cached - main_buffers)
```

Comment:
cAdvisor calculates the memory working set as `MemoryUsed - stats["inactive_files"]` (https://github.com/google/cadvisor/blob/master/container/libcontainer/handler.go#L827-L835).  If we already drop the cached files size here, the result given by cAdvisor is always negative.  Should we just use `mem_used = int64(main_total - main_free)`?

Reply:
Yeah I was just mimicking whatever `free` is showing, not whatever cgroup v1's `memory.usage_in_bytes` is showing (and yes, this was a wrong thing to do). cgroup v1's and v2's (`memory.current` for non-root cgroups) DO include the page cache (which is questionable but that's the way it is).

Thanks for pointing out, fixed.